### PR TITLE
Issue-9: Add initial support for Validation

### LINKED ISSUE #9: Add initial support for Validation
Requires an implementation of `Tuple`.

Must include the following:

* A `Validation` type.
* The following static constructors:
    * `Failure a`
    * `Success b`
* The following static methods:
    * `concat :: Validation a b → Validation a b → Validation a b`: Concatenates two `Failure` values of `a` together, replace a `Success` with the `Failure`; otherwise, take the first `Success.
    * `failures :: [Validation a b] → [a]`: Extracts from a list of `Validation` all of the `Failure` elements in extracted order.
    * `fromFailure :: a → Validation a b → a`: Extracts the value of a `Failure`; otherwise, returns the `defaultValue`.
    * `fromSuccess :: b → Validation a b → b`: Extracts the value of a `Success`; otherwise, returns the `defaultValue`.
    * `isFailure :: Validation a b → Boolean`: Determines whether or not the argument is a `Failure`.
    * `isSuccess :: Validation a b → Boolean`: Determines whether or not the argument is a `Success`.
    * `partitionValidations :: [Validation a b] → ([a], [b])`: Partitions a list of `Validation` into two lists. All `Failure` elements are extracted, in order, to the first position of the output. Similarly for the `Success` elements in the second position.
    * `successes :: [Validation a b] → [b]`: Extracts from a list of `Validation` all of the `Success` elements in extracted order.
    * `validate :: (b → Boolean) → a → b → Validation a b`: Validates a value `b` and returns a `Success` of `b` if the `predicate` returns `true`; otherwise, a `Failure` of `a`.
    * `validationMap :: ([a] → c) → (b → c) → Validation a b → c`: If the value is `Failure a`, apply the first function to `[a]`; otherwise, apply the second function to `b`.

### PR DESCRIPTION
**Pull request by jlmorgan**
- [x] Squash commits before merging.

closes #9
SemVer:Minor
enhancement

### CODE REVIEW

File: src/data/Functions/requireFunction.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+"use strict";
+
+// Project
+const curry = require("../Tuple/curry");
+const isFunction = require("./isFunction");
+const requireType = require("../../requireType");
+
+/**
+ * Throws an {@link TypeError} if the {@code value} provided is not an {@link Function}.
```
Comment (doomchild) [inline src/data/Functions/requireFunction.js]: `an` -> `a`

File: src/data/Validation/Validation.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ */
+class Success extends Validation {
+  constructor(value) {
+    super();
+
+    this._value = value;
+  }
+
+  /**
+   * Determine whether or not the {@code other} has the same value as the current {@code instance}.
```
Comment (doomchild) [inline src/data/Validation/Validation.js]: You have "Determine" here, but "Determines" on other predicates.

File: src/data/Validation/requireValidation.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+"use strict";
+
+// Project
+const { isValidation } = require("./Validation");
+const curry = require("../Tuple/curry");
+const requireType = require("../../requireType");
+
+/**
+ * Throws an {@link TypeError} if the {@code value} provided is not a {@code Validation}.
```
Comment (doomchild) [inline src/data/Validation/requireValidation.js]: `an` -> `a`

File: src/requireType.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+"use strict";
+
+// Project
+const curryN = require("./data/Tuple/curryN");
+
+/**
+ * Throws an {@link TypeError} if the {@code value} provided fails the {@code predicate}.
```
Comment (doomchild) [inline src/requireType.js]: `an` -> `a`

**Review by doomchild — CHANGES_REQUESTED** [review]
Minor documentation formatting stuff is needed.

File: src/data/Functions/requireFunction.js (hunk 1 above)
Comment (jlmorgan) [reply src/data/Functions/requireFunction.js]: Fixed.

File: src/data/Validation/requireValidation.js (hunk 3 above)
Comment (jlmorgan) [reply src/data/Validation/requireValidation.js]: Fixed.

File: src/requireType.js (hunk 4 above)
Comment (jlmorgan) [reply src/requireType.js]: Fixed.

File: src/data/Validation/Validation.js (hunk 2 above)
Comment (jlmorgan) [reply src/data/Validation/Validation.js]: Fixed.

File: src/data/Functions/requireFunction.js (hunk 1 above)
Comment (doomchild) [reply src/data/Functions/requireFunction.js]: Wait, I'm dumb.  This _should_ be `a`.  Sorry.

File: src/data/Validation/requireValidation.js (hunk 3 above)
Comment (doomchild) [reply src/data/Validation/requireValidation.js]: My grammar brain must have been doing something else.  It was correct originally.

File: src/requireType.js (hunk 4 above)
Comment (doomchild) [reply src/requireType.js]: My grammar brain must have been doing something else. It was correct originally.

File: src/data/Functions/requireFunction.js (hunk 1 above)
Comment (jlmorgan) [reply src/data/Functions/requireFunction.js]: I knew what you meant. = )

**Review by doomchild — APPROVED** [review]
Seems legit.

**Pull request merged**